### PR TITLE
DE - Salt Browser

### PR DESCRIPTION
**Pull request by dalejerikson**
This PR represents the first version of the backend routes required for the ACAS Salt Browser to be functional. 

ACAS PR: https://github.com/mcneilco/acas/pull/967

**Status of Each Feature and Sub-Feature**

Query Route 

- Added query to return any salts in the database where the query string matches (LIKE) to any abbrev or names of salts in the database 

Create Salt Route 

- Able to create and register salts successfully 
- Requires JSON salt data which contains a name, abbrev, and mol structure
- Dryrun mode calculates molecular properties but does not save any Salt 
- Validation is done on structure, name, and abbreviation fields being filled 

Edit Salt Route

- Able to change the structure, name, and/or abbrev of any Salt 
- Has a dryrun mode that will return list of warnings/errors including cascading changes and affected dependencies 
- Edit salt route calls restandardization function as well as a dependency check 

Delete Salt Route

- Route uses id of Salt to find; If it exists and there are no lot dependencies then the salt can be deleted 
- Route to delete salt includes validation check to determine if salt is present in any lots; if it is then an error is returned 

Get SDF Route 

- returns SDF string of all salts within database 
- relevant salt properties are included as SDF properties

### CODE REVIEW

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [inline src/main/java/com/labsynch/labseer/api/ApiSaltController.java:171]: Now that I look at this again there is such a thing as isomers so I will want to change this to checking duplicate structures

Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:171]: maybe I will keep this but set the level to warning

File: src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java
Comment (dalejerikson) [inline src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java:90]: Might want to make this catch only CmpdRegMolFormatException

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [inline src/main/java/com/labsynch/labseer/api/ApiSaltController.java:138]: To trim or not to trim...hmmmm.....

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [inline src/main/java/com/labsynch/labseer/api/ApiSaltController.java:144]: I will want to implement a CompareTo method of some sort for salt structs

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (bffrost) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:171]: I like the idea of keeping it but with a warning.

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/api/ApiSaltController.java:142]: Should we be calculating the molecular weight and charge at this point too?

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (bffrost) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:144]: Can we just calculate weight & charge, and skip comparing structures? Structure comparison is a rabbit hole, and we should avoid it if we can.

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (bffrost) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:186]: I answered my own question later in the code review. I did not remember / know about that `standardizationService.restandardizeLots` function which recalculates lot molecular weight.

I'm not quite sure what that has to do with dependent data though. Could you elaborate on the thinking here?

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/api/ApiSaltController.java:231]: I don't think this is the right status code. This seems to have something to do with an "Expect" header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/417

I think a BAD_REQUEST error would be more appropriate here

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/api/ApiSaltController.java:268]: Not sure how to interpret this comment

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/api/ApiSaltController.java:327]: Similar thing here - I think it'd make sense to put the `dryrun` as a RequestParameter

File: src/main/java/com/labsynch/labseer/service/StandardizationService.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/StandardizationService.java:32]: How strange that this function existed but wasn't exposed. Just a nit - the term "restandardize" now has a specific connotation of running chemical structures through a "standardizer" or "preprocessor" or "structure cleaner", which isn't what this function does. It'd be nice to name this something more obvious, like "recalculateLotMolWeights"

File: src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java
Comment (bffrost) [reply src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java:90]: I agree

File: src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java:105]: nit: indentation is a little wonky in this file

File: src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java:127]: We should be careful here. There are two methods to calculate molecular weight: `getMass` which calculates average molecular weight based on isotopic abundance and `getExactMass` which calculates the monoisotopic mass. We're using `getMass` above, so I think this line should use `mol.getMass();` too. `getMass` is sortof our default, then in specific scenarios we may use `getExactMass` when the monisotopic mass is specifically requested.

File: src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java:130]: I'm not sure why these general "catch" statements are here. Is there a reason we don't simply let any errors get thrown and bubble up to whatever code is calling `calculateWeight`? I'm generally opposed to "swallowing" errors unless we have a specific reason for a specific type of error.

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:190]: Brian Bolt did some work recently to refactor and simplify dependency checking. You should have a look at his code and chat with him, and see if there are places where you can reuse functions he's written, or follow that same pattern.

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:233]: This line seems to be looking up Parent compounds whose structure ID matches the structure ID of the salt. This isn't correct.

To find dependencies for Salts, we'll need to find any SaltForms that contain this Salt, then find the dependencies of those SaltForms. We should chat about this.

**Review by bffrost — CHANGES_REQUESTED** [review]
This is a good start to the backend logic. I think it'll help clean up a bunch if you study some of Brian's recent dependency checking code, and then try to reuse as much as that as you can here. Please comment back on any of my feedback that doesn't make sense, and we can work on this also during pair programming.

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:144]: So, just to confirm, there should be no alert if someone makes an isomer of a salt?

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:427]: I don't believe I use `updateFromJSON`. However, I just wanted to keep the ApiSaltController consistent with the updates I was making. I'm confused on the context of what you're saying here.

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:142]: Yes, I will do this since for the weight and charge warning checks

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:231]: Will replace this.

File: src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java:130]: Will remove them or make more specific catches

File: src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/SaltStructureServiceImpl.java:127]: Gotcha. Thanks for the background. Didn't know that :). Will replace this.

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:115]: Yep!

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:118]: Your thoughts are correct. Will do that. Thanks for the suggestion!

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:186]: The thought was if a lot is dependent on a salt's existence and then the salt is updated then the lot should be restandardized and the weights should be updated. I don't have the best grasp of the object relations in ACAS and how Salts, Salt Forms, Lots, Containers, etc. are all connected and related together. Is there a chart or documentation that could help with this?

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:190]: This code was based off of code Brian B. messaged me a few weeks ago but I will follow up and see if it is still representative of what I should be doing.

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:233]: Ah makes sense. Happy to chat about this further. I need to gain a better understanding of identifiers and object relationships in ACAS.

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (bffrost) [reply src/main/java/com/labsynch/labseer/api/ApiSaltController.java:144]: Confirmed. No alert if someone makes an isomer of a salt. We can always add something like that later, but for now let's keep it simple.

**Comment by dalejerikson** [comment]
@bffrost I'd like to review some of the dependency checking and restandardization with you during our scheduled meeting today since the logic is intense and it extends to regions of the system I'm unfamiliar with. Specific cases also include:

~ Checking if the salt abbrev is included in Lot Corp Names
~ If the Lot Corp Name changed, cascade the update to "batch code" values within Analysis Groups and Lot Inventory Containers

**Comment by dalejerikson** [comment]
Rebased Branch

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (dalejerikson) [reply src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:442]: Implemented

**Comment by dalejerikson** [comment]
Backend Ready for Review

File: src/main/java/com/labsynch/labseer/api/ApiSaltController.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/api/ApiSaltController.java:130]: nit: this term restandardization is going to throw people off. Can you rename this error to "ERROR: Hit an issue updating salt dependencies:"

File: src/main/java/com/labsynch/labseer/service/SaltService.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltService.java:15]: nit: I think this function is no longer being used, so it can be removed.

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:211]: This function seems redundant with the much more fleshed out `validateSaltEdit` function. I think you're already doing these two checks within `validateSaltEdit`. Perhaps you could run `validateSaltEdit` then check whether it has any messages with level = `error` to derive your boolean "validSalt" ?

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:393]: nit: you could easily do this with a nested for loop within the `isoSalt` loop you're doing above.
```
Set<Lot> lotSet = new HashSet<Lot>(); 
for (IsoSalt isoSalt : isoSaltsSet){
	SaltForm saltForm = isoSalt.getSaltForm();
	for (SaltForm saltForm : saltFormSet){
		Set<Lot> lots = saltForm.getLots();
		lotSet.addAll(lots);
}
```

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:403]: I think you're missing a check whether the newCorpName is different from the old corp name. If it's not different, then we can skip the calls to `assayService.renameBatchCode` and `containerService.renameBatchCode` and the logging of "new lot corp name"

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:415]: nit: This unused function can be removed to avoid confusion.

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:465]: nit: This unused function can be removed too.

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:491]: nit: This unused function can be removed too.

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:378]: hypernit: you could do this in a one-liner:
```
Collection<IsoSalt> isoSalts =  IsoSalt.findIsoSaltsBySalt(salt).getResultList();
```
This is typically what we do, and I find it a little easier to follow.

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:505]: Same hypernit here

File: src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:511]: same nit here about combining the for loops

Comment (bffrost) [reply src/main/java/com/labsynch/labseer/service/SaltServiceImpl.java:511]: nit: If you really wanted to DRY this up, you might consider a small helper function: `findLotsBySalt` that takes this bit of repeated logic out of both `checkDependentData` and `updateDependencies`

**Review by bffrost — CHANGES_REQUESTED** [review]
@dalejerikson This is getting really close. By and large, the logic all looks right to me. I've noted some leftover / unused functions that you should clean up, and a few other nits, but this will be good to go once you resolve those. Nice work!

**Review by bffrost — APPROVED** [review]
LGTM!!

**Pull request merged**
